### PR TITLE
rework CAN to use esp-idf functionality

### PR DESCRIPTION
**Pull request by the0ne**
only limited testing was posible.
please especially verify that **CAN networking** still works as expected!

### CODE REVIEW

**Comment by the0ne** [comment]
This new CAN implementation turns out to work **really well**.

Is there any test script for testing CAN networking? I got two ESP32-EVB devices with CAN that I could wire together. Anything I need to be aware of with regards to CAN networking?

**Comment by thunderWhitecat** [comment]
@the0ne,

I will focused again on CAN shortly.

**Comment by the0ne** [comment]
fantastic - let me know whenever I can help

File: components/sys/drivers/can.c [hunk 1]
```diff
@@ -65,6 +65,7 @@
 #include <sys/mutex.h>
 
 #include <drivers/can.h>
+#include <drivers/cpu.h>
```
Comment (the0ne) [inline components/sys/drivers/can.c:68]: now required for CPU_FIRST_CAN and CPU_LAST_CAN

**Comment by the0ne** [comment]
after having done quite some testing / using the CAN-bus,
[these](https://github.com/whitecatboard/Lua-RTOS-ESP32/pull/267/commits/ce436fd629e00835c3b7bc1ad081eb19edeb6acd) are the improvements I have implemented.

File: components/sys/drivers/can.c [hunk 2]
```diff
@@ -145,26 +147,41 @@ driver_error_t *can_check_error(esp_err_t error) {
...
+        ; //do nothing
+    } else {
+        can_start();
+        can_initiate_recovery();
+    }
+}
+
 static driver_error_t *can_ll_tx(can_message_t *frame) {
-    can_start(); //make sure can is running
-    can_initiate_recovery(); //make sure the bus is on
```
Comment (the0ne) [inline components/sys/drivers/can.c]: these calls should not be performed on every send so I have moved them to the error-case only

File: components/sys/drivers/can.c [hunk 3]
```diff
@@ -145,26 +147,41 @@ driver_error_t *can_check_error(esp_err_t error) {
...
+        can_recovery();
+        return error;
+    }
     return 0;
 }
 
 static driver_error_t *can_ll_rx(can_message_t *frame) {
-    can_start(); //make sure can is running
-    can_initiate_recovery(); //make sure the bus is on
-
```
Comment (the0ne) [inline components/sys/drivers/can.c]: these calls should not be performed on every send so I have moved them to the error-case only

**Comment by thunderWhitecat** [comment]
Hi Thomas,

I'm working again with CAN on ESP32. Testing this PR with my current CAN nodes HW test, based on PIC18FK MCU doesn't work as expected.

Example code:

-- Attach CAN at 500 Kbps speed
can.attach(can.CAN0, 500000)

-- Send a message
can.send(can.CAN0, 0x200, can.STD, 1, string.pack("B", 0x01))

PIC18FK MCU nodes don't respond to the sended CAN message. May be something related to 500 Kbps?

**Comment by thunderWhitecat** [comment]
Ok, I see the problem.

In current implementation:

https://github.com/whitecatboard/Lua-RTOS-ESP32/wiki/CAN-module

speed is expressed in bits/sec, but in your implementation it's expressed in Kbps.

No problem, I will change the wiki, it's more clear to express speed in Kbps.

Thanks.

**Pull request merged**